### PR TITLE
add types for temp_(v/nmbr/pntr)String, use const

### PR DESCRIPTION
**Pull request by digama0**
The memory management pays a lot of attention to whether a returned `vstring` is to be deallocated by the caller, or it is a temporary value which is automatically deallocated by the next call to `let`. This commit makes it clear which is which by using the return type of the function (where `temp_vstring` is a typedef for `vstring`.)

Similarly, a `vstring` is not actually the same as a C `char *`: a `vstring` can be viewed as a `char *`, but it also has a control block at negative addresses which can be used to reallocate the string. Most functions that take `const vstring` are in fact `const char *` methods, so the type signatures have been changed to reflect this. Similarly, many input arguments can be `const`, which is a C99 thing (which was already in the codebase), so use it more consistently.

(This is one part of a big commit that I'm breaking into pieces for ease of review.)

### CODE REVIEW

**Comment by wlammen** [comment]
Yep.  Necessary.  I already thought of similar measures.

File: mmvstr.h [hunk 1]
```diff
@@ -126,18 +126,26 @@ with 'let(&' and thus has the same effect as 'let(&'.
```
Comment (wlammen) [inline mmvstr.h:126]: I'd like to see a usage example for temp_vstring in the introducing comment block.  Examples should be updated to use vstringdef.

File: mmvstr.c [hunk 2]
```diff
@@ -91,8 +91,7 @@ static void* tempAlloc(long size)  /* String memory allocation/deallocation */
 /* Make string have temporary allocation to be released by next let() */
```
Comment (wlammen) [inline mmvstr.c:91]: How is the connection between makeTempAlloc and temp_vstring?

**Review by wlammen — CHANGES_REQUESTED** [review]

File: mmvstr.c [hunk 3]
```diff
@@ -138,25 +137,23 @@ void let(vstring *target, vstring source) {
...
     }
-    *target= "";
+    *target = "";
   }
 
   freeTempAlloc(); /* Free up temporary strings used in expression computation */
 
 } /* let */
 
-vstring cat(vstring string1,...)        /* String concatenation */
```
Comment (wlammen) [inline mmvstr.c:148]: cat has another limitation not mentioned in the comments of master: the list of sources must not be empty i.e cat(NULL) is illegal

**Review by wlammen — CHANGES_REQUESTED** [review]

File: mmvstr.c [hunk 4]
```diff
@@ -234,50 +225,44 @@ int linput(FILE *stream, const char* ask, vstring *target)
 
 
 /* Find out the length of a string */
-long len(vstring s)
-{
-  return ((long)strlen(s));
+long len(vstring s) {
```
Comment (wlammen) [inline mmvstr.c:228]: is there a particular reason why len does not take a char const* as parameter?

File: mmvstr.c [hunk 5]
```diff
@@ -669,15 +645,13 @@ vstring xlate(vstring sin,vstring table)
 
 
 /* Returns the ascii value of a character */
-long ascii_(vstring c)
-{
+long ascii_(const char *c) {
   return ((long)c[0]);
```
Comment (wlammen) [inline mmvstr.c:649]: The implementation hazardous, because for a non-ASCII character the result depends on the implementation defined signedness of char (signed or unsigned char, you cannot tell).  This is not something that has to do with your PR, but it caught my eye.

**Review by wlammen — CHANGES_REQUESTED** [review]

File: mmdata.c [hunk 6]
```diff
@@ -923,39 +919,39 @@ nmbrString *nmbrCat(nmbrString *string1,...) /* String concatenation */
...
 
 
 
 /* Find out the length of a nmbrString */
-long nmbrLen(nmbrString *s)
+long nmbrLen(const nmbrString *s)
 {
   /* Assume it's been allocated with poolMalloc. */
-  return (((long)(((long *)s)[-1] - (long)(sizeof(nmbrString))))
+  return (((long)(((const long *)s)[-1] - (long)(sizeof(nmbrString))))
```
Comment (wlammen) [inline mmdata.c:945]: I don't think, the cast to long of the nominator is necessary:
((((const long *)s)[-1] - (long)(sizeof(nmbrString)))
is sufficient and still looks complex.
same holds for line 954

File: mmdata.c [hunk 7]
```diff
@@ -1137,8 +1118,7 @@ long nmbrRevInstr(long start_position, nmbrString *string1,
 
 
 /* Converts nmbrString to a vstring with one space between tokens */
-vstring nmbrCvtMToVString(nmbrString *s)
-{
+temp_vstring nmbrCvtMToVString(const nmbrString *s) {
   long i, j, outputLen, mstrLen;
   vstring tmpStr = "";
```
Comment (wlammen) [inline mmdata.c:1123]: Why do you avoid use of vstringdef here?
Suggest to move the declaration of ptr and ptr2 right before the for loop.
The implementation looks spooky if s is empty. In this case space seems to alloc an entry on the tempAllocStack containing the empty string, but where is that freed? let() ignores freeing of empty strings (?).  There seems to be a memory leak.

**Review by wlammen — CHANGES_REQUESTED** [review]

File: mmdata.c (hunk 7 above)
Comment (digama0) [reply mmdata.c:1123]: I'd rather not mess with uses of `vstringdef`, because it will conflict with #38 which does a bulk translation to use the macro everywhere (and it also changes the name from `vstringdef` to `vstring_def` so added references to `vstringdef` in this PR will also cause conflicts). I think if you look you will likely find that this line is touched in #38.

Comment (digama0) [reply mmdata.c:1123]: Same thing about the other suggestions: this PR is about constifying and nothing more (to the extent possible). I have a bunch more changes waiting in the wings but if I put it all in one PR it will be huge and hard to follow, so I'm trying to do one type of change at a time.

Comment (digama0) [reply mmdata.c:1123]: Based on how the code generally handles these temp strings, it seems like empty strings never have to be deallocated. I assume the reason is because `let` will never create an empty string, it always sets the input to some designated `""` in static memory. (If that's not the case, then the whole `let(&x, "")` idiom is suspect, but even beyond that there are parts of the code that say that skip the `let(&x, "")` statement if `!x[0]`, which is bogus if "non-canonical" empty strings can be created in temporary storage.

Comment (digama0) [reply mmdata.c:1123]: Canonicalization of empty strings is done here:

https://github.com/metamath/metamath-exe/blob/ddcd53aa5b6e9403f3692ebde8b42f0dc24aff7f/mmvstr.c#L128-L134

Comment (digama0) [reply mmdata.c:1123]: In this function `nmbrCvtMToVString`, when `s` is empty, it does:
* `mstrLen = 0`
* `outputLen = -1`
* `let(&tmpStr, "")` (so `tempStr` is a pointer to `""` in static storage)
* The `if (tmpStr[0]) makeTempAlloc(tmpStr);` line is skipped
* `return tmpStr`

So it returns a canonical empty `vstring`, which is fine to be treated as a `temp_vstring` because it doesn't need to be deallocated.

File: mmdata.c (hunk 6 above)
Comment (digama0) [reply mmdata.c:945]: There are a lot of redundant casts like this. I think it was being compiled in some mode where the compiler would give a warning if you did not use a cast even for implicit numeric up-casts, but I don't think our current compiler settings enforce this. I think I want to clean this up in a separate PR. (The cast `const long *` from `const nmbrString *` is unnecessary since `nmbrString` is a typedef for `long`.) The casts here look like defense against promotion in the division; probably if we read the C spec it can be simplified a lot.

File: mmdata.c (hunk 7 above)
Comment (wlammen) [reply mmdata.c:1123]: Under the assumption outputLen == -1
in mmdata.c line 1137 nmbrCvtMToVStr calls space(-1)
in mmvstr.c line 571 space(-1) calls string(-1, ' ')
in mmvstr.c line 562 string(0, ' ') calls tempAlloc(1) 
In mmvstr.c line 56: tempAlloc(1) allocates 1 byte memory
Question: when is this byte freed again?
This is important, since you strengthen the return type from vstring to temp_vstring.

Comment (digama0) [reply mmdata.c:1123]: Aha, this is a bug in `string()`, it should return `""` if `n = 0`

Comment (wlammen) [reply mmdata.c:1123]: One should check whether nmbrSeg e. a. suffer from a similar problem.

**Comment by digama0** [comment]
I'd like to get this merged before #38. Are there any more objections? The `string()` bug was not introduced in this PR and will be handled in its own PR.

File: mmdata.c (hunk 7 above)
Comment (digama0) [reply mmdata.c]: On further review (while preparing a separate PR for this issue), I think there is actually no bug here. The invariant is as follows:

* Temporary strings (`temp_vstring`, constructed by `tempAlloc` and freed by `freeTempAlloc`, usually via `let`) are allowed to be noncanonical empty strings, and are also allowed to have internal `\0` characters as a quick truncation operation. They should never be manually freed, because there is no easy means to remove them from the temp stack, and this is why they cannot appear in the left argument of a `let`. Static strings can also be treated as `temp_vstring`, because from client code this only means that the string should not be freed manually and one cannot assume that they will live past the next call to `freeTempAlloc`.

* Regular strings (`vstring`, constructed by `vstring_def`) can be in one of two states: empty, and pointing to the canonical empty string `""`, or nonempty and pointing to a `malloc`'d buffer with a control block at bytes -12..0 relative to the pointer. The `let(&str, value)` function can be used to change the value of a `vstring str`. `vstring` variables always start in the empty state using `vstring_def`, and they must be in the empty state when exiting a scope or else memory will leak.

As it applies to this code, we see that the invariants are upheld:

* `string` has no bug: it may sometimes create noncanonical empty strings but they are always temp-allocated.
* `nmbrCvtMToVString` uses `let(&tmpStr, space(outputLen))` correctly to copy a `temp_vstring` into a `vstring`, and calls `makeTempAlloc` to put it on the temporary stack. It will be deallocated at the next call to `let`.
* `makeTempAlloc` doesn't have a bug per se, but it has IMO the wrong precondition, and one that does not match the corresponding functions `nmbrMakeTempAlloc` and `pntrMakeTempAlloc`. It is supposed to convert a `vstring` into a `temp_vstring` and the way to do that is to put owned (nonempty) `vstring`s on the temp stack for later reclamation, and do nothing for (canonical static) empty `vstring`s since this already qualifies as a `temp_vstring`, it should not be deallocated. Currently, the nonempty check is done outside the function for `makeTempAlloc` and inside the function for `nmbrMakeTempAlloc` and `pntrMakeTempAlloc`; I will fix this in another PR.

File: src/mmdata.c [hunk 8]
```diff
@@ -1137,8 +1118,7 @@ long nmbrRevInstr(long start_position, nmbrString *string1,
 
 
 /* Converts nmbrString to a vstring with one space between tokens */
-vstring nmbrCvtMToVString(nmbrString *s)
-{
+temp_vstring nmbrCvtMToVString(const nmbrString *s) {
   long i, j, outputLen, mstrLen;
   vstring tmpStr = "";
```
[the remark this answers is not in the document]
Comment (digama0) [reply src/mmdata.c:1123]: It is born as a `vstring`, and is only converted to a `temp_vstring` right at the end. I think type ascribing it as a `temp_vstring` from the beginning would be confusing, but there is no real way in C to express a function that changes the type of its argument. How about having `makeTempString` return its argument instead (but with a different type)?

[the remark this answers is not in the document]
Comment (digama0) [reply src/mmdata.c:1123]: Conceptually, you should not think of it as ever bypassing the `makeTempString` call. It is always converted to a temp, it is just that sometimes there is nothing to be done. #46 makes this clearer.

File: src/mmvstr.h [hunk 9]
```diff
@@ -126,18 +126,69 @@ with 'let(&' and thus has the same effect as 'let(&'.
 
 #include <stdio.h>
 
+/* A vstring is like a C string, but it contains a control block allowing
+for memory allocation. New vstrings should always be constructed from the
+`vstringdef` macro. */
 typedef char* vstring;
+
+/* A vstring allocated in temporary storage. These strings will be deallocated
```
Comment (wlammen) [inline src/mmvstr.h:134]: ...vstring is allocated...

File: src/mmdata.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (wlammen) [reply src/mmdata.c:1123]: > How about having makeTempString return its argument instead (but with a different type)?
That's cool.

**Review by wlammen — CHANGES_REQUESTED** [review]

File: src/mmdata.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (wlammen) [reply src/mmdata.c:1123]: Fix the typo, Nothing more.

[the remark this answers is not in the document]
Comment (digama0) [reply src/mmdata.c:1123]: what typo? Are we still talking about this line or about #46?

File: src/mmvstr.h (hunk 9 above)
Comment (digama0) [reply src/mmvstr.h:134]:
```suggestion
/* A vstring is allocated in temporary storage. These strings will be deallocated
```
